client/tinychain/ml/dnn.py

```diff
@@ -64,14 +193,29 @@ def train(self, inputs, cost):
                 A.append(A_l.copy())
                 Z.append(Z_l)
 
-            m = inputs.shape[0]
+            m = inputs.shape[0]   
             dA = cost(A[-1]).sum() / m
 
             updates = []
+            
+            initials=optimizer.create()
```

Comment:
This creates a new optimizer on every training cycle. Surely that's not what you want? For example, how will the time `t` ever be incremented?

Reply:
but its outside of our backward propagation loop so i guess its in correct place?

Reply:
It's in the `train` function which is currently called once on every iteration of the training loop

Reply:
oh i got you  so i think this should be called in splitted function  or just letting user call it first before training but thats not good way to deal with our user

Reply:
umm or how about this / 

in  test_neural_net.py

    def execute(self, cxt):
        @tc.closure
        @tc.post_op
        def while_cond(output: Dense, i: tc.UInt):
            fit = ((output > 0.5) != cxt.labels).any()
            return fit.logical_and(i < MAX_ITERATIONS)
        
        cxt.optimizer = tc.ml.Adam_Optimizer.create()
        @tc.closure
        @tc.post_op
        def train(i: tc.UInt):          
            output = cxt.nn.train(cxt.inputs, lambda output: ((output - cxt.+labels)**2) * LEARNING_RATE , cxt.optimizer)
            return {"output": output, "i": i + 1}

        cxt.result = tc.While(while_cond, train, {"output": cxt.nn.eval(cxt.inputs), "i": 0})

        response = self.host.post(ENDPOINT, cxt)
        self.assertLess(response["i"], MAX_ITERATIONS, "failed to converge")

Reply:
I like how this looks but I'm still unclear on how the state of the optimizer will be updated

Reply:
we have a loop on our training function and inside the training function we call  .optimize method and .update method 
 what is wrong with that?

Reply:
That's fine but the `optimize` method returns the new state of the optimizer, so how does that new state make it to the next iteration? We can discuss this over chat if it's still unclear

Reply:
i got you now , umm i guess we need better idea

Reply:
how about this ? 
in dnn.py


        def train(self, inputs, cost,optimizer):
            '''
            train function will take the inputs and cost function
            and give back last layer output after activation.
            
            A : list of Tensors starts with Inputs Tensor the Result of each layer after activation.
            Z : list of Output Tensors before getting applied with activation function.
            
            m : number of data rows
            
            cost : is a cost function which takes the output and actual output as input
                 and calculate the difference (e.g. : lambda output:((output- actual_labels)**2)* learning_Rate)
                 
           optimizer : is your selected optimizer to tune your Neural Network learning paramters.           
            
           updates : list of updated values for weights and biases.
           dA : Average Delta Error 
           A[i]: Previous Layer Tensor.
           Z[i+1] : Current Layer Tensor.
           A[-1]: Last Layer after activation ( Prediction ).
           After(updates, A[-1]) : Output of Last Layer after applying Activation and updating weights.
            '''

            A = [inputs]
            Z = [None]

            for layer in layers:
                A_l, Z_l = layer.train_eval(A[-1])
                A.append(A_l.copy())
                Z.append(Z_l)

            m = inputs.shape[0]   
            dA = cost(A[-1]).sum() / m

            updates = []
            

            l_m,l_v,t=optimizer['l_m'],optimizer['l_v'],optimizer['t']
            
            for i in reversed(range(0, num_layers)):
                
                updated_values = optimizer.optimize(A[i],dA,Z[i+1],decay_rate_1=0.9,
                                   decay_rate_2=0.99,epsilon=10e-8 ,l_m ,l_v ,t)
                l_m,l_v,t=updated_values['l_m'],updated_values['l_v'],updated_values['t']
                d_weights,d_bias=updated_values['d_weights'],updated_values['d_bias']               
                update= optimizer.update(d_weights,d_bias)
                updates.append(update)
                
            optimizer_state = l_m,l_v,t
               

            return After(updates, A[-1]) , optimizer_state

Reply:
and in test_neural_net.py 

    def execute(self, cxt):
        @tc.closure
        @tc.post_op
        def while_cond(output: Dense, i: tc.UInt):
            fit = ((output > 0.5) != cxt.labels).any()
            return fit.logical_and(i < MAX_ITERATIONS)
        
        cxt.optimizer = tc.ml.Adam_Optimizer.create()
        @tc.closure
        @tc.post_op
        def train(i: tc.UInt):          
            output,cxt.optimizer = cxt.nn.train(cxt.inputs, lambda output: ((output - cxt.+labels)**2) * LEARNING_RATE , cxt.optimizer)
            return {"output": output, "i": i + 1}

        cxt.result = tc.While(while_cond, train, {"output": cxt.nn.eval(cxt.inputs), "i": 0})

        response = self.host.post(ENDPOINT, cxt)
        self.assertLess(response["i"], MAX_ITERATIONS, "failed to converge")

Reply:
You'll have to include the optimizer state in the state of the While loop:
```
    cxt.optimizer = tc.ml.Adam_Optimizer.create()
    
    @tc.closure
    @tc.post_op
    def train(i: tc.UInt):          
        output, optimizer = cxt.nn.train(cxt.inputs, lambda output: ((output - cxt.+labels)**2) * LEARNING_RATE , cxt.optimizer).unpack(2)
        return {"output": output, "optimizer": optimizer, "i": i + 1}

    cxt.result = tc.While(while_cond, train, {"output": cxt.nn.eval(cxt.inputs), "i": 0, "optimizer": cxt.optimizer})
```

Reply:
I tried testing two layers but the training didn't converge within the 30s timeout. I'm expecting AdamOptimizer to improve the training times substantially.

Reply:
i have 2 questions 

output,optimizer = cxt.nn.train(cxt.inputs, lambda output: ((output - cxt.+labels)**2) * LEARNING_RATE , cxt.optimizer)


what is  `cxt.+labels` ? 
and why not assigning to  cxt.optimizer ? and u used optimizer instead ?

Reply:
i guess its just type mistake ?

Reply:
`cxt.optimizer` means "the state in the current op context called 'optimizer'" so it can't be reset to a new state. You can try running the code to see what I mean.

The code that protects `Context` states from being reassigned is in Context.__setattr__: https://github.com/haydnv/tinychain/blob/main/client/tinychain/util.py#L59 so if you want to see how your program works without it you can try taking out the "if name in self.form: raise ValueError..." check